File: Algorithm/C++/1000+/1471.the-k-strongest-values-in-an-array.cpp
/*
 * @lc app=leetcode id=1471 lang=cpp
 *
 * [1471] The k Strongest Values in an Array
 */

// @lc code=start
#include <vector>
#include <algorithm>

class Solution {
public:
    std::vector<int> getStrongest(std::vector<int>& arr, int k) {
        std::sort(arr.begin(), arr.end());
        int mid = arr[int((arr.size()-1)/2)];

        int left = 0, right = arr.size() - 1;
        std::vector<int> result;
        result.reserve(k);
        while (k > 0) {
            if (std::abs(arr[right] - mid) >= std::abs(arr[left] - mid)) {
                result.push_back(arr[right]);
                right--;
            } else {
                result.push_back(arr[left]);
                left++;
            }
            k--;
        }
        return result;
    }
};
// @lc code=end

File: Algorithm/Python/1000+/1471.the-k-strongest-values-in-an-array.py
#
# @lc app=leetcode id=1471 lang=python3
#
# [1471] The k Strongest Values in an Array
#

# @lc code=start
        
class Solution:
    def getStrongest(self, arr: List[int], k: int) -> List[int]:
        arr.sort()
        n = len(arr)
        mid = arr[int((n-1)/2)]
        
        left, right, result = 0, n-1, []
        while k > 0:
            if abs(arr[right] - mid) >= abs(arr[left] - mid):
                result.append(arr[right])
                right -= 1
            else:
                result.append(arr[left])
                left += 1
            k -= 1
        return result
    
# @lc code=end

